tket/src/Converters/CoherentTableauConverters.cpp

```diff
@@ -0,0 +1,672 @@
+    if (rten.second.string.map.size() != 0) {
+      // Reached the rows with non-empty output segment
+      break;
+    }
+    // Else, we add the row to the subspace
+    to_diag.push_back({rten.first, 1.});
+  }
+  unsigned post_selected_size = to_diag.size();
+  std::set<Qubit> diag_ins{input_qubits.begin(), input_qubits.end()};
+  Circuit in_circ = mutual_diagonalise(to_diag, diag_ins, CXConfigType::Tree);
+  // Extract the dagger of each gate in order from tab
+  for (const Command& com : in_circ) {
+    auto args = com.get_args();
+    qubit_vector_t qbs = {args.begin(), args.end()};
+    tab.apply_gate(
+        com.get_op_ptr()->dagger()->get_type(), qbs,
+        CoherentTableau::TableauSegment::Input);
+  }
+
+  // Diagonalised rows should still be at the bottom of the tableau - reduce
+  // them to a minimal set of qubits for post-selection by first reducing to
+  // upper echelon form
+  std::vector<std::pair<unsigned, unsigned>> row_ops =
+      gaussian_elimination_row_ops(
+          tab.tab_.zmat_.bottomRows(post_selected_size));
+  for (const std::pair<unsigned, unsigned>& op : row_ops) {
+    tab.tab_.row_mult(op.first, op.second);
+  }
+  // Obtain CX instructions as column operations
+  std::vector<std::pair<unsigned, unsigned>> col_ops =
+      gaussian_elimination_col_ops(
+          tab.tab_.zmat_.bottomRows(post_selected_size));
+  // These gates will also swap qubits to isolate the post-selections on the
+  // first few qubits - this is fine as can be cleaned up later with peephole
+  // optimisations
+  for (const std::pair<unsigned, unsigned>& op : col_ops) {
+    tab.tab_.apply_CX(op.first, op.second);
+    CoherentTableau::col_key_t ctrl = tab.col_index_.right.at(op.first);
+    CoherentTableau::col_key_t trgt = tab.col_index_.right.at(op.second);
+    in_circ.add_op<Qubit>(OpType::CX, {ctrl.first, trgt.first});
+  }
+
+  // Post-select rows
+  // Post-selection op is not yet available in tket - replace this once
+  // implemented
+  if (post_selected_size != 0)
+    throw std::logic_error(
+        "Not yet implemented: post-selection required during CoherentTableau "
+        "synthesis");
+  // for (unsigned r = 0; r < post_selected_size; ++r) {
+  //   CoherentTableau::row_tensor_t row = tab.get_row(tab.get_n_rows() - 1);
+  //   if (row.second.string.map.size() != 0 || row.first.string.map.size() != 1
+  //   || row.first.string.map.begin()->second != Pauli::Z) throw
+  //   std::logic_error("Unexpected error during post-selection identification
+  //   in CoherentTableau synthesis"); Qubit post_selected_qb =
+  //   row.first.string.map.begin()->first; if (row.second.coeff == -1.)
+  //   in_circ.add_op<Qubit>(OpType::X, {post_selected_qb});
+  //   tab.remove_row(tab.get_n_rows() - 1);
+  //   post_selected.insert(post_selected_qb);
+  //   throw std::logic_error("Not yet implemented: post-selection required
+  //   during CoherentTableau synthesis");
+  // }
```

Comment:
Assume this is left intentionally as an implementation hint?

Reply:
Yep, will make this more clear in the preceding comment